pynicotine/gtkgui/search.py

```diff
@@ -280,6 +286,77 @@ def file_search_response(self, msg):
 
         page.file_search_response(msg)
 
+    # Download Folder
+    def _folder_contents_response(self, msg, check_num_files=True):
+        """ Peer code: 37 """
+        """ When we got a contents of a folder, get all the files in it, but
+        skip the files in subfolders """
+
+        username = msg.init.target_user
+        file_list = msg.list
+ 
+        log.add_transfer("Received response for folder content request from user %s", username)
+
+        folderFiles = {
+            "user": username,
+            "virtualpath": None,
+            "destination": None,
+            "filenames" : []             
+        }
+
+        for i in file_list:
+            for directory in file_list[i]:
+                if os.path.commonprefix([i, directory]) != directory:
+                    continue
+
+                files = file_list[i][directory][:]
+                num_files = len(files)
+
+                if check_num_files and num_files > 100:
+                    events.emit("download-large-folder", username, directory, num_files, msg)
+                    return
+
+                destination = core.transfers.get_folder_destination(username, directory)
+                folderFiles["destination"] = destination
+
+                virtualpath = directory.rstrip("\\") + "\\"
+                folderFiles["virtualpath"] = virtualpath
+
+                if num_files > 1:
+                    files.sort(key=lambda x: strxfrm(x[1]))
+
+                log.add_transfer(("Attempting to download files in folder %(folder)s for user %(user)s. "
+                                  "Destination path: %(destination)s"), {
+                    "folder": directory,
+                    "user": username,
+                    "destination": destination
+                })
+
+                for file in files:
+                    size = file[2]
+                    h_bitrate, _bitrate, h_length, _length = slskmessages.FileListMessage.parse_result_bitrate_length(
+                        size, file[4])
+
+                    folderFiles["filenames"].append({
+                        "filename": file[1],
+                        "size": file[2],
+                        "h_bitrate": h_bitrate,
+                        "h_length": h_length
+                    })
+                    # Put Dialog To Select the Files    
+
+
+                    # Then Call this so get the File (it does checks here etc)
+                    # Will need to move this somewhere else
+                    # core.transfers.get_file(username, virtualpath, path=destination, size=size, bitrate=h_bitrate, length=h_length)
+                    
+        if self.download_folder is None:
+            self.download_folder = DownloadFolder(self.window.application)
+
+        self.download_folder.set_folder_list(folderFiles)
+        self.download_folder.show()                
```

Comment:
I think it would be better to immediately show the dialog when pressing "Download Folder", and add all the visible files to core.transfers.requested_folders. Adding to my previous review comment, when the rest of the files are sent, you could listen to the `folder-contents-response` event in downloadfolder.py, and add the rest of the files from core.transfers.requested_folders to the listview.

Reply:
Ideally, we would also show a message if the user is offline and the full file list can't be retrieved, but that's something I could look into later.

Reply:
OK, whew!
I have done this. 

The Dialog will now show instantly on selecting download folder and populate upon receiving the folder content event response (will show 'loading' in info_bar while waiting).

Once response is received (either success or error) the dialog will be populated or show error message.
The dialog will also handle the > 100 files returned in folder.

On completing this change it seems logical now that the Folder Dialog can manage all the behaviour to with downloading a folder from the search tab. I have subsequently removed all references to the download folder from search, transfers and downloads. Please check this change acceptable, I do feel however it is cleaner and easier to follow now.

Reply:
Thanks!

> The dialog will also handle the > 100 files returned in folder.

This can be removed. It was only in place since blindly downloading a folder could sometimes result in a lot of files.

Reply:
ok